### ISSUE TITLE
Random authentication error when using account SAS to upload large file to blob storage

### ISSUE THREAD

**Issue by thanhnhut90** [Azure/azure-storage-python#297]
Hi guys,

I am seeing random authentication issues when using account SAS to upload large files ( >20 GB) to blob storage.
```
azure.common.AzureHttpError: Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:aa7f238b-0001-0006-3d71-cf3f37000000
Time:2017-05-18T00:54:58.0436290Z</Message><AuthenticationErrorDetail>sr is mandatory. Cannot be empty</AuthenticationErrorDetail></Error>
```
The authentication error detail showed `sr is mandatory. Cannot be empty`. Per [Account SAS documentation](https://docs.microsoft.com/en-us/azure/storage/storage-dotnet-shared-access-signature-part-1), account SAS doesn't have the `sr` parameter while service SAS has it. Using the same token, I was able to successfully upload 2 large files at different time, but it also failed many times due to authentication issue. The SAS token is given to me, and I do not have direct access to the account, so I may not be able to change or generate a service SAS.
Could you please take a look if this is a bug or simply I just miss something?

Thanks

**Comment by jofriedm-msft** [comment]
Hi @thanhnhut90,

The error code does appear misleading but comparing this request with one that succeeded around the same time, I see that this failure appears to be because the SAS token was duplicated in the uri.

failure:

'sv=2016-05-31&ss=b&srt=sco&sp=rwdlac&se=2037-05-10T20:07:37Z&st=2017-05-10T12:07:37Z&spr=https&sig=XXXXX&sv=2016-05-31&ss=b&srt=sco&sp=rwdlac&se=2037-05-10T20:07:37Z&st=2017-05-10T12:07:37Z&spr=https&sig=XXXXX&comp=block&blockid=TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01UQTRORFkwTnpBeE5EUSUzRA%3D%3D' 

successful:

sv=2016-05-31&ss=b&srt=sco&sp=rwdlac&se=2037-05-10T20:07:37Z&st=2017-05-10T12:07:37Z&spr=https&sig=XXXXX&comp=block&blockid=TURBd01EQXdNREF3TURBd01EQXdNREF3TURBd01URXdOamczTmpneU5UWSUzRA%3D%3D

**Comment by thanhnhut90** [comment]
Hello @jofriedm-msft. Thank you so much for looking into this. As far as I can tell, I don't see duplication in my code (the token is hard-coded and wasn't changed when it successfully uploaded or failed to upload), but I copy the code below in case if I really miss something.
```
def upload(service, path):
    '''upload to azure'''
    start_time = int(time.time())
    print "Start time: {}".format(start_time)
    print "File size {} bytes".format(os.stat(path).st_size)
    try:
        data = create_blob_path(path)
        service.create_blob_from_path(data['container', data['blob'], path, progress_callback=progress_callback, validate_content=True, max_connections=5)
        print service.make_blob_url(data['container'], data['blob'])
        end_time = int(time.time())
        print "End time: {}".format(end_time)
        print "Total {} mins".format(float(end_time - start_time)/60)
    except InvalidPathException as error:
        print str(error)

def progress_callback(current, total):
    percent = current/total
    if (percent % 10 == 0) and (percent != 0):
        print current
        print total
    	print "percent: {}".format(current/total)
	print "================================="

def delete_blob(service, container, blob):
    '''delete a blob from comtainer'''
    try:
        service.delete_blob(container, blob)
        print "Delete {} successfully".format(blob)
        return
    except Exception as error:
        print "Unable to delete the blob due to {}".format(str(error))

def list_blobs_in_container(service, container):
    '''list all blobs in a container'''
    blobs = service.list_blobs(container)
    for blob in blobs:
        print blob.name

def main():
    bblock_service = BlockBlobService(account_name=XXXXXX,\
                                        sas_token=XXXXXXX)
    args = usage()
    path = args.get('path', None)
    list_container = args.get('list', None)
    blob_name = args.get('delete', None)
    if path:
        upload(bblock_service, path)
    elif list_container:
        list_blobs_in_container(bblock_service, 'somecontainer')
    elif blob_name:
        delete_blob(bblock_service, 'somecontainer', blob_name)
    else:
        print "do nothing because no argument is given"
```

**Comment by jofriedm-msft** [comment]
@thanhnhut90 , the below line is cut off for me but it looks like you are forgetting a ']' to close out the data['container']. If that doesn't fix it are you seeing this for all requests are just upload because that can narrow down your investigation, and maybe paste the rest of that line 😎 

service.create_blob_from_path(data['container', data['blob'], path, progress_callback=progress_callback

**Comment by thanhnhut90** [comment]
actually it was my failed attempt to hide some information :). In the test code, I simply hard-coded a container, so it looks more like this
`service.create_blob_from_path('somecontainer', data['blob'], path, progress_callback=progress_callback, validate_content=True, max_connections=5)`

today i tried 3 times. 2 success and 1 failure; however, I made a modification to the SAS token by appending `&sr=c` (this was before I opened the issue), and the error I got back was different as expected

Thu May 18 08:29:10 PDT 2017 -> failed. Exception was thrown 10 mins after the script was run.
```
RequestId:08dc8852-0001-0056-45ec-cf203f000000
Time:2017-05-18T15:39:54.8875260Z</Message><AuthenticationErrorDetail>Signature fields not well formed.</AuthenticationErrorDetail></Error>
```
Thu May 18 09:26:41 PDT 2017 -> success (same token as above)
Thu May 18 10:14:35 PDT 2017 -> success (same token as above)

**Comment by thanhnhut90** [comment]
Last Friday (May 19, 2017) I was able to upload around 10 large files without getting any exception. Today, I tried to upload from multiple servers. Again, same code, same token.
For large files (30 GB +): got exception for all of them
For smaller files (100-ish MB or less): no exception
```
#Attempt 1:
<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthenticationFailed</Code><Message>Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:f942c126-0001-010a-4058-d39793000000
Time:2017-05-23T00:09:43.0639278Z</Message><AuthenticationErrorDetail>sr is mandatory. Cannot be empty</AuthenticationErrorDetail></Error>
-------------------------------------------------------------------------------------------------
#Attempt 2:
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
InsecurePlatformWarning
Start time: 1495498791 (Tue May 23 00:19:51 UTC 2017)
File size 66485960 bytes (57 MB)
End time: 1495498808 (Tue May 23 00:20:08 UTC 2017)
Total 0.283333333333 mins
```

**Comment by rambo-msft** [comment]
Thanks for the additional information @thanhnhut90 .
It's not immediately clear why this is occurring intermittently, I will run some tests on my end to try and reproduce these scenarios, but it's likely directly related to uploading larger payloads.

**Comment by neh** [comment]
I'm having the same intermittent failure with uploading large files, although most of my files are a little smaller. I've seen this failure occur with files as small as 2-5 GB.

**Comment by rambo-msft** [comment]
Thanks @neh.
Just an update, this appears to be an issue with out retry policies when using SAS (which will often get triggered with bigger uploads). We are looking into getting a fix out soon.

**Comment by thanhnhut90** [comment]
@rambo-msft: do you have any update on this issue? 
Thank for looking into this :)

**Comment by matthchr** [comment]
@rambo-msft I am also seeing this issue very intermittently with files around 2GB

**Comment by jofriedm-msft** [comment]
Hi all, sorry for the delayed response. I have found the issue and we will get a fix out for this soon. The issue appears to only occur when SAS is used and a retry is performed. As a temporary workaround you can use SharedKey or disable retries.

**Comment by huguesv** [comment]
Did you miss this PR?
https://github.com/Azure/azure-storage-python/pull/299

**Comment by matthchr** [comment]
@jofriedm-msft - How is disabling retries a workaround? If a retry occurring is what triggers the problem if I disable retries won't I just fail due to the error that triggered the retry (instead of due to this error)?

**Comment by jofriedm-msft** [comment]
@huguesv yes, I didn't see that someone had already sent a PR for this.

@matthchr, yeah I guess it depends on what error handling you have built on top of the library. I was just thinking that disabling retries would get you back the actual failure reason instead of the misleading "sr is mandatory" failure but using shared key authentication is the real workaround.

**Comment by zezha-msft** [comment]
We have fixed this and will make a release very shortly.

**Comment by zezha-msft** [comment]
The fix has been released with 0.34.3. Thanks for your patience. 😃

**Comment by Tower450** [comment]
Getting the same error for no reason the code was working well for like a month.
